### ISSUE TITLE
dotnet.exe silently ignores global.json with a two-digit sdk version

### ISSUE THREAD

**Issue by maltalex** [dotnet/runtime#96299]
### Describe the bug

dotnet.exe silently ignores global.json with a "wrong" sdk version like "6.0" or "3.1" (as opposed to "6.0.0" or "3.1.0").

### To Reproduce
Install .Net SDK, create an empty directory with a global.json file with a two digit SDK version:

```
{
  "sdk": {
    "version": "6.0"
  }
}
```

Now, run `dotnet --info` in the directory and observe output similar to the following:

```
.NET SDK:
 Version:           8.0.100
 Commit:            57efcf1350
 Workload version:  8.0.100-manifests.6c33ef20

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100\

.NET workloads installed:
 Workload version: 8.0.100-manifests.6c33ef20
There are no installed workloads to display.

Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71

...

global.json file:
  Not found
```

Note that dotnet reports that `global.json` hasn't been found. Commands like `dotnet --version` or `dotnet build` will **silently** use the latest SDK instead of requiring .NET 6 as was intended by global.json. The expected behavior in this case would be to either treat "6.0" as "6.0.0" or to show an error to point out the invalid version number.

**Comment by ShreyasJejurkar** [comment]
> If global.json specifies "version": "9.0" and you have versions 9.0.100 and 9.0.200 installed, then which one should it run

Yeah, this is correct, I missed this totally. I am not sure why we start from `.100` patch version instead of going from `0`,. 

But anyway, I guess it would be good if we just showed an error if the version specified it's just 2 digits.

**Comment by svick** [comment]
@KalleOlaviNiemitalo It's actually a way to have four-part version number inside a three-part version number. So 9.0.100 is effectively 9.0.1.0. What is somewhat funny is that the second part will always be zero (but that's only the case since .Net Core 3.1).

**Comment by elinor-fung** [comment]
Related:
- https://github.com/dotnet/runtime/issues/94496

As in the above, I don't think it is worth making a breaking change to make it fail instead of ignore the global.json. But updating the output of `--info` to indicate that it was found but invalid / ignored for whatever reason would be good - similar to https://github.com/dotnet/runtime/issues/94496#issuecomment-1800339684.

**Comment by maltalex** [comment]
@elinor-fung I'm sorry, but that's a bad solution. No one actively runs `--info` after updating their `global.json` to check whether the change worked. You'll just be preserving the current state of **silent** failure, which is the worst possible outcome.

**Comment by elinor-fung** [comment]
Any change would be in a component used by all .NET versions, so I wouldn't want the breaking/active failure.

I wouldn't be opposed to relaxing the version parsing to make a missing part imply 0 (instead of the current strict semver 2 requiring all major, minor, and patch - https://semver.org/#spec-item-2) - but that still seems insufficient per the concerns @KalleOlaviNiemitalo [raised](https://github.com/dotnet/runtime/issues/96299#issuecomment-1868601999).

**Comment by maltalex** [comment]
@elinor-fung The breaking failure would be only in cases which are already broken, it's just that the failure is silent, and therefore dangerous.

**Comment by elinor-fung** [comment]
Oh, I think my concern about the shared/global component might actually address the concern with relaxing it in the latest version only. I believe the SDK just uses that shared `hostfxr` component to determine SDK resolution based on global.json - @nagilson would you be able to confirm?

If that's the case, then saying missing implies 0 seems reasonable. And the defaults for not specifying things like 'rollForward' remains the same - that is, 6.0 with no rollForward specified -> 6.0.0 with no rollForward specified -> 6.0.0 defaulting to latest patch.

**Comment by gusarov** [comment]
@elinor-fung can you elaborate a bit more about what is valid and what is invalid global json? If I malform the json syntax, it will not be ignored, it will fail to build. I wonder why in some cases global.json considered malformed and ignored in a first place instead of indicating this right away? In our case we did not knew that 3 digits is mandatory, so we put 2 and everything went fine because we had all .Net versions installed anyway and the latest was silently picked up. It ended up with broken CI, but all of this could have been avoided if "dotnet build" either properly failed the build or imply 3 digits out of 2

**Comment by vlad2135** [comment]
Interesting, this comment says that dotnet build ignores broken global.json - https://github.com/dotnet/runtime/issues/94496#issuecomment-1800339666 , but in my experiment with .NET SDK 8.0.100 shows that dotnet build crashes spectacularly.

**Comment by elinor-fung** [comment]
Sorry, this fell off my radar. Looking at it again, it seems we have two slightly different definitions of invalid `global.json`:
- malformed: comments are allowed, but otherwise must conform to JSON RFC
- missing/unexpected data: based on what is expected to be in a `global.json` (the two-digit version falls into this category)

Previously, both these cases were treated the same way (ignored). Some time after https://github.com/dotnet/runtime/issues/94496#issuecomment-1800339680, most SDK commands started failing on malformed global.json.

> the failure is silent, and therefore dangerous.

Agreed - thanks for pushing on this. I did not fully grasp the unsafe nature of the current behaviour - especially since having an ignored global.json currently means latest version including prereleases, which also has security implications. I don't know the original reasoning for the current behaviour, but I agree it is mis-designed from an experience and a reliability/security view - we should take a breaking change to make both malformed and missing/unexpected data an error.

**Comment by elinor-fung** [comment]
Related in terms of security implications of the current behaviour to use latest SDK including prerelease: https://github.com/dotnet/sdk/issues/49777
cc @richlander
